lua/nvim-tree/git/init.lua

```diff
@@ -143,23 +143,33 @@ function M.load_project_status(cwd)
     timeout = M.config.git.timeout,
   }
 
+  local git_directory = git_utils.get_git_directory(project_root)
+
   local watcher = nil
   if M.config.filesystem_watchers.enable then
     log.line("watcher", "git start")
 
-    local callback = function(w)
-      log.line("watcher", "git event scheduled '%s'", w.project_root)
-      utils.debounce("git:watcher:" .. w.project_root, M.config.filesystem_watchers.debounce_delay, function()
-        if w.destroyed then
-          return
-        end
-        reload_tree_at(w.project_root)
-      end)
-    end
+    if git_directory == nil then
+      log.line("watcher", "could not found the location of .git folder")
+    else
+      local callback = function(w)
+        log.line("watcher", "git event scheduled '%s'", w.project_root)
+        utils.debounce("git:watcher:" .. w.project_root, M.config.filesystem_watchers.debounce_delay, function()
+          if w.destroyed then
+            return
+          end
+          reload_tree_at(w.project_root)
+        end)
+      end
 
-    watcher = Watcher:new(utils.path_join { project_root, ".git" }, WATCHED_FILES, callback, {
-      project_root = project_root,
-    })
+      -- Add $GIT_DIR to the list of directory to ignore
+      local base_gitdir = utils.path_basename(git_directory)
+      table.insert(M.config.filesystem_watchers.ignore_dirs, base_gitdir)
```

Comment:
> This is, however, a hack, not a solution.

I agree.

IMHO, the issue boils down to `create_watcher` being responsible to know what can be watched or not. I would argue it's the caller's responsibility.

## Proposal

- add `watch.disable_watcher(path)`
   - Adds to a new list `M.ignored_paths` which is initialized with `IGNORED_PATHS`
- update `create_watcher` to not check if the node is ignored. ([remove this part](https://github.com/nvim-tree/nvim-tree.lua/blob/5860ac3950252384af440847fc532367f0f91aee/lua/nvim-tree/explorer/watch.lua#L43-L45))
- update all callers to check first if the node is in `M.ignored_paths` before calling `create_watcher`

What's your thought?

Reply:
> However, `create_watcher` still have too much responsibility to me.

That is sensible. `is_git` is doing too much, as well as being broken.

Reply:
> the caller check if the node is part of the config

That could be tricky, as there are many call sites.

Adding to the ignored list as per https://github.com/nvim-tree/nvim-tree.lua/pull/2012/files#diff-6e591951880a9ff855ceca8fb5a150c85deeb19593cd5143ed8b19a125f0f9e7R166 does seem to be the right way to go. The git project is loaded before enumerating the tree so the git directory will be in the `ignore_dirs`. I'll document that assumption.

Filtering `IGNORED_PATHS` and `ignore_dirs` is thus Good Enough. The current solution does that.

Let's be safe and not mutate config:
1. `watch.lua` setup copies `filesystem_watchers.ignore_dirs` into `M.ignore_dirs` 
2. create `M.ignore_dir(path)` in `watch.lua` that adds to `M.ignore_dirs`
3. `load_project_status` calls `watch.ignore_dir(base_gitdir)`

Reply:
Testing in a variety of situations (mostly flipping around directories, reloading) shows that `load_project_status` is (always?) run before creating regular filesystem watchers. This will prevent any unwanted watchers.

Even if we do end up creating extra watchers, it's not such a big drama; we are already doing that when the git directory is opened.